lib/fog/core/cache.rb

```diff
@@ -0,0 +1,210 @@
+  #       puts "could not find any cache data for security groups on #{compute}"
+  #       get_resources_and_dump
+  #
+  # == Extending cache backends
+  #
+  # Currently this is on-disk using yml. If need be, this could be extended to other cache backends:
+  #
+  # Find references of yaml in this file, split out to strategy objects/diff backends etc.
+  class Cache
+
+    # cache associated not found
+    class CacheNotFound < StandardError; end
+
+    # cache directory problem
+    class CacheDir < StandardError; end
+
+    # where different caches per +service+ api keys, regions etc, are stored
+    # see the +sandbox=+ method.
+    SANDBOX_PREFIX = "~/.fog-cache"
+
+    # when a resource is used such as `server.cache.dump` the model klass is passed in
+    # so that it can be identified from a different session.
+    attr_reader :model
+
+    # Loads cache associated to the +model_klass+ and +service+ into memory.
+    #
+    # If no cache is found, it will raise an error for handling:
+    #
+    #   rescue Fog::Cache::CacheNotFound
+    #     set_initial_cache
+    #
+    def self.load(model_klass, service)
+      cache_files = Dir.glob("#{namespace(model_klass, service)}/*")
+
+      raise CacheNotFound if cache_files.empty?
+
+      loaded = cache_files.map do |path|
+        model_klass = Object.const_get(load_cache(path)[:model_klass])
+        model_klass.new(load_cache(path)[:attrs])
+      end
+
+      collection_klass = load_cache(cache_files.sample)[:collection_klass] &&
+                         Object.const_get(load_cache(cache_files.sample)[:collection_klass])
+
+      loaded.each do |i|
+        # See https://github.com/fog/fog-aws/issues/354#issuecomment-286789702
+        i.collection = collection_klass.new(:service => service) if collection_klass
+        i.instance_variable_set(:@service, service)
+      end
+
+      # uniqe-ify based on the total of attributes. duplicate cache can exist due to
+      # `model#identity` not being unique. but if all attributes match, they are unique
+      # and shouldn't be loaded again.
+      uniq_loaded = loaded.uniq { |i| i.attributes }
+      if uniq_loaded.size != loaded.size
+        Fog::Logger.warning("Found duplicate items in the cache. Expire all & refresh cache soon.")
+      end
+
+      uniq_loaded
+    end
+
+    # creates on-disk cache of this specific +model_klass+ and +@service+
+    def self.create_namespace(model_klass, service)
+      FileUtils.mkdir_p(self.namespace(model_klass, service))
+    end
+
+    # Expires cache - this does not expire all cache associated.
+    # Instead, this will remove all on-disk cache of this specific +model_klass+ and and +@service+
+    def self.expire_cache!(model_klass, service)
+      FileUtils.rm_rf(namespace(model_klass, service))
+    end
+
+    # loads yml cache from path on disk
+    def self.load_cache(path)
+      @cache ||= {}
+      return @cache[path] if @cache[path]
+      @cache[path] = YAML.load(File.read(path))
```

Comment:
I just realized this. 

Since you are saving the result on a instance variable in the class (not into the object) and every time you call this method you save more data into that instance variable we might have some performance issues here regarding memory, no?

Is the content being saved here going to be too large? Should we have concerns about this?